### PR TITLE
Ability to toggle full text indexing

### PR DESCRIPTION
**Pull request by danielmarbach**
- [x] Additional tests for mixed data
- [x] Additional tests for recoverability search by body

### CODE REVIEW

**Review by SzymonPobiega — APPROVED** [review]

File: src/ServiceControl.AcceptanceTests/Recoverability/MessageFailures/When_failed_message_searched_by_body_content.cs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+    class When_failed_message_searched_by_body_content : AcceptanceTest
+    {
+        [Test]
+        public async Task Should_be_found_when_fulltext_search_enabled()
+        {
+            // setting it even if it is the default
+            SetSettings = settings => settings.EnableFullTextSearchOnBodies = true;
+
+            var searchString = "42";
```
Comment (mikeminutillo) [inline src/ServiceControl.AcceptanceTests/Recoverability/MessageFailures/When_failed_message_searched_by_body_content.cs:22]: In theory, it's possible that the string `42` may appear in a guid and therefore this test (and the one below) could randomly fail. Maybe changing it `forty-two` :)

File: src/ServiceControl.Audit.AcceptanceTests/Auditing/When_processed_message_searched_by_body_content.cs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+    class When_processed_message_searched_by_body_content : AcceptanceTest
+    {
+        [Test]
+        public async Task Should_be_found_when_fulltext_search_enabled()
+        {
+            // setting it even if it is the default
+            SetSettings = settings => settings.EnableFullTextSearchOnBodies = true;
+
+            var searchString = "42";
```
Comment (mikeminutillo) [inline src/ServiceControl.Audit.AcceptanceTests/Auditing/When_processed_message_searched_by_body_content.cs:21]: Same potential issue as the test against the error instance. It's possible that any Guid could contain the string `42` and cause this test to be wonky

**Comment by mikeminutillo** [comment]
A couple of other random thoughts that came from reviewing https://github.com/Particular/ServiceControl/pull/2501

- Both of these PRs add some kind of document store initialize property to enable tests but the mechanism for calling it is in different places. We need to make sure it doesn't get called twice.
- Both PRs add a new user-visible setting. They should probably get added to the startup diagnostics. [This is how the other PR does ie](https://github.com/Particular/ServiceControl/pull/2501/files?file-filters%5B%5D=.cs#diff-b862e8242aa6c06cfffc9b029ce1214b3fbb055a9a3ff6b2f1a7eb64c64fb6f4R255)

File: src/ServiceControl/CompositeViews/Messages/MessagesBodyTransformer.cs [hunk 3]
```diff
@@ -1,21 +1,25 @@
...
     {
         public MessagesBodyTransformer()
         {
             TransformResults = messages => from message in messages
-                                           let metadata =
-                                               message.ProcessingAttempts != null
-                                                   ? message.ProcessingAttempts.Last().MessageMetadata
-                                                   : message.MessageMetadata
+                                           let metadata = message.ProcessingAttempts != null
+                                               ? message.ProcessingAttempts.Last().MessageMetadata
```
[the remark this answers is not in the document]
Comment (danielmarbach) [reply src/ServiceControl/CompositeViews/Messages/MessagesBodyTransformer.cs:14]: Yep we do that all the time like this so I sticked to it

File: src/ServiceControl.AcceptanceTests/Recoverability/MessageFailures/When_failed_message_searched_by_body_content.cs (hunk 1 above)
Comment (danielmarbach) [reply src/ServiceControl.AcceptanceTests/Recoverability/MessageFailures/When_failed_message_searched_by_body_content.cs:22]: Addressed

File: src/ServiceControl.Audit.AcceptanceTests/Auditing/When_processed_message_searched_by_body_content.cs (hunk 2 above)
Comment (danielmarbach) [reply src/ServiceControl.Audit.AcceptanceTests/Auditing/When_processed_message_searched_by_body_content.cs:21]: Addressed

File: src/ServiceControl.UnitTests/BodyStorage/BodyStorageEnricherTests.cs [hunk 4]
```diff
@@ -19,48 +21,83 @@ public class BodyStorageEnricherTests
...
+
+            await enricher.StoreErrorMessageBody(body, attempt);
 
             Assert.AreEqual(body, metadata["Body"], "Body should be stored if below threshold");
+            Assert.IsNull(attempt.Body, "Body property was set but shouldn't have been");
+            Assert.AreEqual(0, fakeStorage.StoredBodySize);
+        }
+
+        [Test]
+        public async Task Should_store_body_in_metadata_when_full_text_disabled_and_not_binary_and_below_LOH_threshold()
```
[the remark this answers is not in the document]
Comment (danielmarbach) [reply src/ServiceControl.UnitTests/BodyStorage/BodyStorageEnricherTests.cs:62]: Addressed

**Comment by danielmarbach** [comment]
When upgrading

image

When upgrading and splitting

image

and if you select NO then the config value is populated to the split settings

image

which you can override

**Comment by danielmarbach** [comment]
Edit view

image

the only drawback it has is that it is also shown on older instances that are not yet upgrade. Right now we have no concept of conditionally enable certain settings on the UI

File: src/ServiceControl.Audit.AcceptanceTests/Auditing/When_processed_message_searched_by_body_content.cs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        public async Task Should_not_be_found_when_fulltext_search_disabled()
+        {
+            SetSettings = settings => settings.EnableFullTextSearchOnBodies = false;
+
+            var searchString = "forty-two";
+
+            var context = await Define<MyContext>()
+                .WithEndpoint<Sender>(b => b.When((bus, c) => bus.Send(new MyMessage
+                {
+                    Something = "Somewhere in the body is the answer to all of the questions. forty-two"
```
Comment (SzymonPobiega) [inline src/ServiceControl.Audit.AcceptanceTests/Auditing/When_processed_message_searched_by_body_content.cs:58]: I love it ❤️

File: src/ServiceControl.UnitTests/BodyStorage/BodyStorageEnricherTests.cs [hunk 6]
```diff
@@ -19,48 +21,83 @@ public class BodyStorageEnricherTests
         public async Task Should_store_body_in_storage_when_binary_and_below_LOH_threshold()
```
[the remark this answers is not in the document]
Comment (SzymonPobiega) [reply src/ServiceControl.UnitTests/BodyStorage/BodyStorageEnricherTests.cs:21]: The On/Off applies only to below threshold, non-binary so we are actually introducing here only one new state.

**Review by SzymonPobiega — APPROVED** [review]

**Comment by mikeminutillo** [comment]
I added a little more detail to the upgrade message as customers may not know if they want full text indexing or not.

image

This makes it clearer that the feature is related to the message search function in ServiceInsight, and that it is having a performance impact on the instance.

I think we should consider reversing the question too, so rather than

> Do you want to continue to support full text indexing for messages?

We should swap it to:

> Do you want to disable full text indexing for messages?

**Comment by danielmarbach** [comment]
@mikeminutillo I like the idea of swapping the message 👍

**Comment by mikeminutillo** [comment]
I think there's an issue with the powershell flag as it stands. At the moment, the flag is optional and defaults to true, so it will enable full text indexing if no value is supplied, _even if a prior upgrade had disabled it_. I think we should switch it to `DisableFullTextSearchOnBodies`. If it's set, then we can change the config value to `false` and turn off the feature. If it is not set, then we do nothing, and the instance should get the default value of `true`.

**Comment by danielmarbach** [comment]
@mike or how about `FullTextSearchOnBodies`?

**Comment by danielmarbach** [comment]
Or maybe make the default a nullable bool?

**Comment by mikeminutillo** [comment]
I think that the upgrade is about transitioning to the new feature (which is the ability to turn off full text indexing) and not about being able to control which features are on/off. Other features with a toggle do not appear in the upgrade questsions. We typically assume that the prior value was correct. In this case, the options are:

a) Turn it off
b) (The default) Leave it as it was (which is on)

**Review by tmasternak — APPROVED** [review]

**Pull request merged**